### DISCUSSION TITLE
How to use System.Configuration.ConfigurationManager with Uno

### DISCUSSION THREAD

**Discussion by trydalch** [unoplatform/uno#17689] [Q&A] [answered]
We have a bunch of legacy .net framework code that targets 4.7.2 or 4.8. Because of .Net Compatibility Mode (see https://learn.microsoft.com/en-us/dotnet/core/porting/#net-framework-compatibility-mode), it appears to work with Uno when we build for a windows app (packaged and unpackaged. But when we run it on an android emulator, we see a bunch of these errors: 

```
[monodroid-assembly] open_from_bundles: failed to load assembly System.Configuration.ConfigurationManager.dll
Loaded assembly: /data/data/com.companyname.TestDLLAndroid3/files/.__override__/System.Configuration.ConfigurationManager.dll [External]
```

(note, we tried with different projects as part of our debugging and got the same kinds of errors). 

Even though we added System.Configuration.ConfigurationManager, created an app.config and made sure to have it copy to the output directory, we were unable to get any values from it in even the most trivial example. 

Our question is this: Is it possible to use our .NET Framework DLLs (ours and Microsoft's from Nuget like

![6a7a5568-06e3-46f1-9649-ed73a60cbb52](https://github.com/user-attachments/assets/5f1d9520-87a3-4212-a51f-3f8dbe3d950a)
 System.Configuration.ConfigurationManager) with Uno 

![cf56ba93-255f-4459-87c7-f58bf5ef3a83](https://github.com/user-attachments/assets/4315ce07-73bf-4a75-9a38-5e552eb2089c)
when building for Android (and by extension, we'd want to know about iOS, too).

[msbuild.zip](https://github.com/user-attachments/files/16366541/msbuild.zip)

**Comment by jeromelaban** [comment]
Thanks for the question. It's theoretically possible to use those netfx assemblies, but it highly depends on how what they use, and how those are implemented on mobile platforms.

While it's not a Uno-specific topic, it would be easier if you were providing a repro sample. Also, your binlog file did not upload.

**Reply by trydalch** [reply]
Thanks Jerome. I uploaded my binlog. 

This is the most simple example: In a fresh project with the System.Configuration.ConfigurationManager nuget package installed, I modified the MainViewModel.cs to be:

```
namespace TestDLLAndroid3.Presentation;

public partial class MainViewModel : ObservableObject
{
    private INavigator _navigator;

    [ObservableProperty]
    private string? name;

    public MainViewModel(
        IStringLocalizer localizer,
        IOptions<AppConfig> appInfo,
        INavigator navigator)
    {
        _navigator = navigator;
        Title = "Main";
        Title += $" - {localizer["ApplicationName"]}";
        Title += $" - {appInfo?.Value?.Environment}";
        GoToSecond = new AsyncRelayCommand(GoToSecondView);
    }
    public string? Title { get; }

    public ICommand GoToSecond { get; }

    private async Task GoToSecondView()
    {
        var configAppSettings = System.Configuration.ConfigurationManager.AppSettings;
        Console.WriteLine(configAppSettings.ToString());
        //await _navigator.NavigateViewModelAsync<SecondViewModel>(this, data: new Entity(Name!));
    }

}
```

and created an app.config file. When I run the app targeting windows, it loads the app.config file contents. When I run it as android, it's null.

**Reply by jeromelaban** [reply]
Got it. The file system is different on Android, depending on how you place your file. Setting the file as `Content` in the csproj is not doing the same as it does on Windows or desktop targets, and the file will not end up at the same place because of the way android is setup. Same goes for iOS, Catalyst and WebAssembly.

You may need to programmatically place the file in a known location (like the documents folder, or similar) then indicate to the ConfigurationManager on where to find it.

**Reply by trydalch** [reply]
Ah. That would explain it. Do you have any guidance on figuring out where it's currently placed? Or where I can look to figure it out?

**Reply by jeromelaban** [reply]
On Android files placed as content are saved as android assets and have no standard "File.IO" presence. You need to use https://platform.uno/docs/articles/features/file-management.html#support-for-storagefilegetfilefromapplicationuriasync to get them.

**Reply by trydalch** [reply]
Alright, I've tried accessing the file as shown there in the docs. Is the ms-appx:// scheme standard? Or is it different for each app? I'm getting a folder not found when I just try to get the root folder.

**Reply by trydalch** [reply]
Ah, I see the Microsoft docs now. Those are standard schemas.

**Reply by trydalch** [reply, marked as answer]
I've got it solved. The steps are this:

1. Set the app.config Build Action to Content
2. Copy the config file to the ApplicationData folder
3. Set the APP_CONFIG_FILE environment variable
4. Force a config refresh

Here's are the essential lines of code:

```c#
// Get the config file from the app package
StorageFile sourceFile = await StorageFile.GetFileFromApplicationUriAsync(new Uri("ms-appx:///app.config"));

// Get the destination folder (LocalFolder in this case)
StorageFolder destinationFolder = ApplicationData.Current.LocalFolder;

// Copy the file to the destination folder
destinationFile = await sourceFile.CopyAsync(destinationFolder, "app.config", NameCollisionOption.ReplaceExisting);

// Set the copied file as the current config file 
AppDomain.CurrentDomain.SetData("APP_CONFIG_FILE", destinationFile.Path);

// Refresh the app settings section to tell ConfigManager to load from the file
System.Configuration.ConfigurationManager.RefreshSection("appSettings");
```

**Reply by trydalch** [reply]
Thank you @jeromelaban!

**Reply by trydalch** [reply]
Here's the solution for iOS 

``` 
// Get the source file from the app bundle
var sourceFilePath = Foundation.NSBundle.MainBundle.PathForResource("app", "config");
if (sourceFilePath == null)
{
	throw new FileNotFoundException("app.config not found in bundle.");
}

// Get the destination folder (LocalFolder in this case)
var destinationFolder = Environment.GetFolderPath(Environment.SpecialFolder.Personal);
var destinationFilePath = Path.Combine(destinationFolder, "app.config");

// Copy the file to the destination folder
File.Copy(sourceFilePath, destinationFilePath, true);

// Set the configuration path
AppDomain.CurrentDomain.SetData("APP_CONFIG_FILE", destinationFilePath);

// Refresh the app settings section to tell ConfigManager to load from the file
System.Configuration.ConfigurationManager.RefreshSection("appSettings");
```
